src/memory/sqlite_store.rs

```diff
@@ -0,0 +1,225 @@
+    }
+
+    /// Connects to a known DB url using the given URL.
+    pub fn connect(url: &str, backoff: Option<bool>) -> Result<Self, SqlMemoryError> {
+        // SqliteConnectionManager::file is the equivalent of using
+        // `rusqlite::Connection::open` as documented in the function's source
+        // code.
+        let pool = r2d2::Pool::builder()
+            .max_size(MAX_POOL_SIZE)
+            .build(SqliteConnectionManager::file(url))?;
+        pool.get().unwrap().execute(CREATE_TABLE_SCRIPT, [])?;
+
+        Ok(Self {
+            pool,
+            exponential_backoff: backoff.unwrap_or(false),
+            _marker: PhantomData,
+        })
+    }
+}
+
+impl<const ADDRESS_LENGTH: usize, const WORD_LENGTH: usize> MemoryADT
+    for SqlMemory<Address<ADDRESS_LENGTH>, [u8; WORD_LENGTH]>
+{
+    type Address = Address<ADDRESS_LENGTH>;
+    type Error = SqlMemoryError;
+    type Word = [u8; WORD_LENGTH];
+
+    async fn batch_read(
+        &self,
+        addresses: Vec<Self::Address>,
+    ) -> Result<Vec<Option<Self::Word>>, Self::Error> {
+        if addresses.is_empty() {
+            return Ok(Vec::new());
+        }
+        let mut conn = self.pool.get()?;
+
+        // Using a transaction to ensure the read is atomic
+        let tx = conn.transaction()?;
+
+        let query = format!(
+            "SELECT adr, word FROM findex_aw_store WHERE adr IN ({})",
+            (0..addresses.len())
+                .map(|_| "?")
+                .collect::<Vec<&str>>()
+                .join(",")
+        );
+
+        // Convert addresses to SQL parameters
+        let params: Vec<ToSqlOutput> = addresses
+            .iter()
+            .map(|addr| {
+                rusqlite::types::ToSqlOutput::Borrowed(rusqlite::types::ValueRef::Blob(&**addr))
+            })
+            .collect();
+
+        // Execute the query
+        let mut stmt = tx.prepare(&query)?;
+        let mut rows = stmt.query(rusqlite::params_from_iter(params))?;
+
+        // Process all rows
+        let mut result_map = std::collections::HashMap::new();
+        while let Some(row) = rows.next()? {
+            let adr: Self::Address = row.get::<_, [u8; ADDRESS_LENGTH]>(0)?.into();
+            let word: Self::Word = row.get(1)?; // tb verified
+            // Convert to fixed-size array
+            result_map.insert(adr, word);
+        }
+
+        // Create results in the same order as input addresses
+        // TODO: Check if this is necessary of if the order is always respected anyway
+        let results = addresses
+            .iter()
+            .map(|addr| result_map.get(addr).copied())
+            .collect();
+
+        Ok(results)
+    }
+
+    async fn guarded_write(
+        &self,
+        guard: (Self::Address, Option<Self::Word>),
+        bindings: Vec<(Self::Address, Self::Word)>,
+    ) -> Result<Option<Self::Word>, Self::Error> {
+        let mut conn = self.pool.get()?;
+
+        let (guard_address, guard_value) = guard;
```

Comment:
This is not a value but a word.
```suggestion
        let (ag, wg) = guard;
```